File: HanJaehee/programmers/43163/programmers43163.java
public class Solution {
    
    int answer = Integer.MAX_VALUE;
    boolean[] v;
    public int solution(String begin, String target, String[] words) {
        v = new boolean[words.length];
        dfs(begin, target, words, 0);
        if(answer == Integer.MAX_VALUE) return 0;
        return answer;
    }
    
    public void dfs(String now, String target, String[] words, int count){
        if(now.equals(target)){
            answer = Math.min(answer, count);
            return;
        }
        
        for(int i=0; i<words.length; ++i){
            if(!v[i] && check(now, words[i])){
                v[i] = true;
                dfs(words[i], target, words, count+1);
                v[i] = false;
            }
        }
    }
    
    public boolean check(String now, String target){
        int count = 0;
        for(int i=0; i<now.length(); ++i){
            if(now.charAt(i) != target.charAt(i)) count++;
        }
        if(count == 1) return true;
        return false;
    }
     
}

File: swanious/programmers/43163/programmers_43163.py
import sys

answer = sys.maxsize

def checkDiffOneWord(str1, str2):
    cnt = 0
    for i in range(len(str1)):
        if str1[i] != str2[i]:
            cnt += 1
    return cnt == 1


def dfs(start, count, target, visit, words):
    global answer

    if start == target:
        answer = min(count, answer)
        return True

    for i in range(len(words)):
        if not visit[i] and checkDiffOneWord(start, words[i]):
            visit[i] = True
            if dfs(words[i], count + 1, target, visit, words): return True
            visit[i] = False
            count -= 1

    return False


def solution(begin, target, words):
    l = len(words)
    visit = [False] * len(words)
    
    return 0 if not dfs(begin, 0, target, visit, words) else answer